### PR TITLE
api-server: door health

### PR DESCRIPTION
**Pull request by koonpeng**
## What's new

This adds a watchdog for doors, currently the only check is a heartbeat on the door state.

There is also a change in logging/database schema, now we only keep the working set in the database, logs are printed to stdout to be collected by another service. Presumably the logs will be collected in a separate database optimized for writes, so the sql database will just be used for the current/latest states.

## Self-checks

- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

**Comment by matiasbavera** [comment]
> There is also a change in logging/database schema, now we only keep the working set in the database, logs are printed to stdout to be collected by another service. Presumably the logs will be collected in a separate database optimized for writes, so the sql database will just be used for the current/latest states.

Cool! We can read from a file or send them via HTTP request to fluentd, I'll ask for your opinion on the other PR =)

File: packages/api-server/api_server/rmf_io/gateway.py [hunk 1]
```diff
@@ -4,18 +4,28 @@
...
-        self.building_map = BehaviorSubject(initial_building_map)
-
+        # NOTE: the rx type hints don't actually work https://github.com/ReactiveX/RxPY/issues/514
+        self.door_states: Subject[DoorState] = Subject()
+        self.current_door_states: Dict[str, DoorState] = {}
         self.door_states.subscribe(self._update_door_states)
 
+        self.door_health: Subject[DoorHealth] = Subject()
+        self.current_door_health: Dict[str, DoorHealth] = {}
+        self.door_health.subscribe(self._update_door_health)
```
Comment (matiasbavera) [inline packages/api-server/api_server/rmf_io/gateway.py:21]: QQ here, sorry, just to see if I'm getting it right, Subject() Represents an object that is both an observable sequence as well as an observer. Each notification is broadcasted to all subscribed observers.

and then you're subscribing an observer to the observable sequence. By using a callback.

and then after instantiating the rmf_gateway, you're using it in the `rmf_io` pushing another observer to the observable sequence again `self.rmf_gateway.door_states.subscribe(self._on_door_state)` .

Can I say that the rmf_gateway is a Singleton?

**Review by matiasbavera — APPROVED** [review]
LGTM, thanks for this, I like this structure better.  Left a couple of comments. 

Should we have a similar heartbeat for the alarm too?

File: packages/api-server/api_server/app.py [hunk 2]
```diff
@@ -31,9 +31,29 @@ async def init_tortoise():
...
 
+async def load_states(repo: RmfRepository, gateway: RmfGateway):
+    logger.info("loading states from database...")
+
+    door_states = await repo.read_door_states()
+    for state in await repo.read_door_states():
+        gateway.door_states.on_next(state)
+    logger.info(f"loaded {len(door_states)} door states")
+
+    logger.info("successfully loaded all states")
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/app.py:54]: nah, I just messed up.

File: packages/api-server/api_server/models/door_state.py [hunk 3]
```diff
@@ -4,21 +4,22 @@
...
 
 
 class DoorState(Model):
-    # TODO: foreign key with Door
-    door_name = fields.CharField(255)
-    current_mode = fields.IntEnumField(DoorModeEnum)
+    # KP: We could have a FK to Door,
+    # but I decided against it to make things easier (pesky race conditions).
+    # If we receive a door name that does not exist in the building,
+    # just take it at face value and write into the database regardless.
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/models/door_state.py:13]: do you mean to make `door_name` a nullable FK? A door state will always contains a door_name though, but the problem is that the door_name may not be registered yet because rmf does not have a enrollment phase. So there is a race condition if we receive the door state before the updated building map.

File: packages/api-server/api_server/rmf_io/operators/heartbeat.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+from rx.operators import buffer_with_time_or_count, distinct_until_changed
+from rx.operators import map as rx_map
+
+
+def heartbeat(liveliness):
+    """
+    Projects a source observable sequence to a boolean. The resulting value is True when there
+    is an observable emitted within the liveliness factor and False otherwise.
+    Args:
+        liveliness: The amount of time between each event for the observable to be considered
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/rmf_io/operators/heartbeat.py:11]: done

File: packages/api-server/api_server/rmf_io/gateway.py (hunk 1 above)
Comment (koonpeng) [reply packages/api-server/api_server/rmf_io/gateway.py:21]: > and then after instantiating the rmf_gateway, you're using it in the rmf_io pushing another observer to the observable sequence again self.rmf_gateway.door_states.subscribe(self._on_door_state) .

The observer subscribes to the observable sequence. There are "producers" like `RmfTransport` and `HealthWatchdog` that pushes items into the subject, then there are "consumers" like `rmf_io` that subscribes to these subjects and perform some tasks accordingly.

> Can I say that the rmf_gateway is a Singleton?

It probably will be in practice, but isn't an actual singleton in implementation.

File: packages/api-server/api_server/models/door_state.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/api-server/api_server/models/door_state.py:13]: I meant keeping the `door_name` but adding another prop that is called `door_fk` to improve the query performance in the future. But I get now and yeah, you're right.
